src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleExceptionShouldNotInitializeToNullOrZero = new(
+        id: DiagnosticIds.PreferDeclarationInitializersExceptNullOrZero,
+        title: "Should not initialize to null or zero",
+        messageFormat: "Do not initialize to null or zero as these already occur by default",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Info,
+        isEnabledByDefault: true,
+        description: "Field initialization to null or zero is redundant and should be avoided.",
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleExceptionShouldNotInitializeInDeclaration = new(
+        id: DiagnosticIds.PreferDeclarationInitializersExceptWhenVaryingInitializations,
+        title: "Should not initialize in declaration due to diverging initializations in constructors",
+        messageFormat: "Do not initialize a field in its declaration if you have diverging initializations in constructors. This is to prevent unnecessary allocations.",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Info,
+        isEnabledByDefault: true,
+        description: "Field initialization should not occur when there are diverging initializations in constructos. This is to prevent unnecessary allocations.",
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleShouldInitializeInDeclarationWhenNoInitializationPresent = new(
+        id: DiagnosticIds.PreferDeclarationInitializersWhenNoInitializationPresent,
+        title: "Should initialize in declaration when no initialization present",
+        messageFormat: "Initialize the field in its declaration when no distint initializations will occur in constructors",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Info,
+        isEnabledByDefault: true,
+        description: "Field initialization should occur in the declaration unless there are diverging initializations in constructors, or the field is a value type or nullable being initialized to the default.",
+        helpLinkUri: HelpLinkUri);
+
+    /// <inheritdoc />
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GeneralRule, RuleExceptionShouldNotInitializeToNullOrZero, RuleExceptionShouldNotInitializeInDeclaration, RuleShouldInitializeInDeclarationWhenNoInitializationPresent);
+
+    /// <inheritdoc />
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.ClassDeclaration);
+    }
+
+    private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
```

Comment:
Consider short-circuit logic and early returns. If you look at other analyzers, it's littered with checks to bail out. We want to defer or noop out of work if we can. Right now, looks like methods like `AnalyzeNode` and `HandleArgumentsList` process every node fully before deciding whether to skip or continue

Reply:
I'm changing it so we iterate through the child nodes in the constructor declaration once, and only based on the type of node. Then, I am caching the field declarations in a list (sadly another mem alloc), and reusing that shorter list rather than going through all child nodes again

Reply:
Update: I forgot to switch to ChildNodes in the PropertyDeclarations. I will address that in the next commit